pkg/uefivars/boot/efiDppMedia.go

```diff
@@ -0,0 +1,258 @@
+		log.Printf("Sig Type 1: unimplemented/cannot identify")
+	} else {
+		// no sig, would need to compare partition #/start/len
+		log.Printf("Sig Type %d: unimplemented/cannot identify", e.SigType)
+		return nil, ErrNotFound
+	}
+	if len(blocks) != 1 {
+		log.Printf("blocks: %#v", blocks)
+		return nil, ErrNotFound
+	}
+	return &HddResolver{BlockDev: blocks[0]}, nil
+}
+
+//return the partition table type as a string
+func (e *DppMediaHDD) pttype() string {
+	switch e.PartFmt {
+	case 1:
+		return "MBR"
+	case 2:
+		return "GPT"
+	default:
+		return "UNKNOWN"
+	}
+}
+
+//return the signature as a string
+func (e *DppMediaHDD) sig() string {
+	switch e.SigType {
+	case 1: //32-bit MBR sig
+		return fmt.Sprintf("%x", binary.LittleEndian.Uint32(e.PartSig[:4]))
+	case 2: //GUID
+		return e.PartSig.ToStdEnc().String()
+	default:
+		return "(NO SIG)"
+	}
+}
+
+//struct in EfiDevicePathProtocol for DppMTypeFilePath.
+//if multiple are included in a load option, concatenate them.
+type DppMediaFilePath struct {
+	Hdr EfiDevicePathProtocolHdr
+
+	PathNameDecoded string //stored as utf16
+}
+
+var _ EfiDevicePathProtocol = (*DppMediaFilePath)(nil)
+
+func ParseDppMediaFilePath(h EfiDevicePathProtocolHdr, b []byte) (*DppMediaFilePath, error) {
+	if len(b) < int(h.Length)-4 {
+		return nil, ErrParse
+	}
+	path, err := uefivars.DecodeUTF16(b[:h.Length-4])
+	if err != nil {
+		return nil, err
+	}
+	//remove null termination byte, replace windows slashes
+	path = strings.TrimSuffix(path, "\000")
+	path = strings.Replace(path, "\\", string(os.PathSeparator), -1)
+	fp := &DppMediaFilePath{
+		Hdr:             h,
+		PathNameDecoded: path,
+	}
+	return fp, nil
+}
+
+func (e *DppMediaFilePath) Header() EfiDevicePathProtocolHdr { return e.Hdr }
+func (e *DppMediaFilePath) ProtoSubTypeStr() string {
+	return EfiDppMediaSubType(e.Hdr.ProtoSubType).String()
+}
+func (e *DppMediaFilePath) String() string {
+	return fmt.Sprintf("File(%s)", e.PathNameDecoded)
+}
+
+func (e *DppMediaFilePath) Resolver() (EfiPathSegmentResolver, error) {
+	fp.Clean(e.PathNameDecoded)
+	pr := PathResolver(e.PathNameDecoded)
+	return &pr, nil
+}
+
+//struct in EfiDevicePathProtocol for DppMTypePIWGFV
+type DppMediaPIWGFV struct {
+	Hdr EfiDevicePathProtocolHdr
+	Fv  []byte
+}
+
+var _ EfiDevicePathProtocol = (*DppMediaPIWGFV)(nil)
+
+func ParseDppMediaPIWGFV(h EfiDevicePathProtocolHdr, b []byte) (*DppMediaPIWGFV, error) {
+	if h.Length != 20 {
+		return nil, ErrParse
+	}
+	fv := &DppMediaPIWGFV{
+		Hdr: h,
+		Fv:  b,
+	}
+	return fv, nil
+}
+func (e *DppMediaPIWGFV) Header() EfiDevicePathProtocolHdr { return e.Hdr }
+func (e *DppMediaPIWGFV) ProtoSubTypeStr() string {
+	return EfiDppMediaSubType(e.Hdr.ProtoSubType).String()
+}
+func (e *DppMediaPIWGFV) String() string {
+	var g uefivars.MixedGuid
+	copy(g[:], e.Fv)
+	return fmt.Sprintf("Fv(%s)", g.ToStdEnc().String())
+}
+
+func (e *DppMediaPIWGFV) Resolver() (EfiPathSegmentResolver, error) {
+	return nil, ErrUnimpl
```

Comment:
Could I suggest -- I know this kind of cuts out a lot of work you've done, but -- if there are types you leave untested and the `Resolver` isn't implemented, is it really useful yet? I'm wondering if we shouldn't remove some of these and leave them unimplemented until we can do that part justice. 

If you want to keep them because they're otherwise useful, please do! Just want you to think about it.

(what is a PIWGFV? I don't even know what it means lol)

Reply:
IIRC `PIWG` is some UEFI/DMTF/? working group. `FV` is firmware volume, i.e. a filesystem embedded in the firmware. The similarly-named `DppMediaPIWGFF` is a file on such a firmware volume.

I do think it's useful without every one of them being implemented - the "device path protocol" is used for more than boot entries, and my understanding is that only a subset can be used for booting.

For my purposes, I only care about resolving boot entries pointing to binaries on "normal" storage media. If someone has a use for other types and can provide a qemu command line (vm config) and qemu firmware file to test against, I'd be willing to help. But I think that should be separate PR(s).

Reply:
sg.